crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_none_literal.rs

```diff
@@ -36,35 +40,39 @@ use crate::{checkers::ast::Checker, settings::types::PythonVersion};
 /// ## Fix safety and availability
 /// This rule's fix is marked as safe unless the literal contains comments.
 ///
-/// There is currently no fix available if there are other elements in the `Literal` slice aside
-/// from `None` and [`target-version`] is set to Python 3.9 or lower, as the fix always uses the
-/// `|` syntax to create unions rather than `typing.Union`, and the `|` syntax for unions was added
-/// in Python 3.10.
+/// There is currently no fix available when applying the fix would lead to
+/// a `TypeError` from an expression of the form `None | None` or when we
+/// are unable to import the symbol `typing.Optional` and the Python version
+/// is 3.9 or below.
 ///
 /// ## References
 /// - [Typing documentation: Legal parameters for `Literal` at type check time](https://typing.readthedocs.io/en/latest/spec/literal.html#legal-parameters-for-literal-at-type-check-time)
 #[derive(ViolationMetadata)]
 pub(crate) struct RedundantNoneLiteral {
-    other_literal_elements_seen: bool,
+    union_kind: UnionKind,
 }
 
 impl Violation for RedundantNoneLiteral {
     const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
 
     #[derive_message_formats]
     fn message(&self) -> String {
-        if self.other_literal_elements_seen {
-            "`Literal[None, ...]` can be replaced with `Literal[...] | None`".to_string()
-        } else {
-            "`Literal[None]` can be replaced with `None`".to_string()
+        match self.union_kind {
+            UnionKind::NoUnion => "`Literal[None]` can be replaced with `None`".to_string(),
+            UnionKind::TypingOptional => {
+                "`Literal[None, ...]` can be replaced with `Optional[Literal[...]]`".to_string()
+            }
+            UnionKind::BitOr => {
+                "`Literal[None, ...]` can be replaced with `Literal[...] | None`".to_string()
```

Comment:
While we're here, it might be nice to change these so they're a little more imperative:

```suggestion
            UnionKind::NoUnion => "Use `None` rather than `Literal[None]`".to_string(),
            UnionKind::TypingOptional => {
                "Use `Optional[Literal[...]]` rather than `Literal[None, ...]`".to_string()
            }
            UnionKind::BitOr => {
                "Use `Literal[...] | None` rather than `Literal[None, ...]`".to_string()
```